From 343c3ca9731e56ba1cafb7df873f14f77ac387e3 Mon Sep 17 00:00:00 2001
From: David Runge <dave@sleepmap.de>
Date: Fri, 28 Jul 2023 01:45:37 +0200
Subject: [PATCH] Fix bootchooser for the output of efibootmgr 18

Since efibootmgr 18, the default output of `efibootmgr` is now more
verbose [1], which breaks the assumptions made by RAUC in regards to
parsing the output. This issue is also affecting others [2].

Fix the parsing logic by unconditionally splitting on a tab in the
detected name part of the boot entry, so that the unused data part can
be discarded.
Emit a debug message for each found boot entry, as this helps in
detecting issues in the future.

[1] https://github.com/rhboot/efibootmgr/commit/8ec3e9dedb3cb62f19847794012420b90f475398
[2] https://github.com/rhboot/efibootmgr/issues/169

Signed-off-by: David Runge <dave@sleepmap.de>
---
 src/bootchooser.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/bootchooser.c b/src/bootchooser.c
index 679c93977..4807e4ffd 100644
--- a/src/bootchooser.c
+++ b/src/bootchooser.c
@@ -1159,11 +1159,26 @@ static gboolean efi_bootorder_get(GList **bootorder_entries, GList **all_entries
 	}
 
 	while (g_match_info_matches(match)) {
+		gchar **name_data_splits = NULL;
+		gchar *name_and_data = g_match_info_fetch(match, 2);
+		name_data_splits = g_strsplit(name_and_data, "\t", 0);
+		if (g_strv_length (name_data_splits) == 0) {
+			g_set_error(
+					error,
+					R_BOOTCHOOSER_ERROR,
+					R_BOOTCHOOSER_ERROR_FAILED,
+					"Extracting a boot entry name failed!");
+			res = FALSE;
+			goto out;
+		}
+
 		efi_bootentry *entry = g_new0(efi_bootentry, 1);
 		entry->num = g_match_info_fetch(match, 1);
-		entry->name = g_match_info_fetch(match, 2);
+		entry->name = g_strdup(name_data_splits[0]);
 		entries = g_list_append(entries, entry);
 		g_match_info_next(match, NULL);
+		g_debug("Detected boot entry %s: %s", entry->num, entry->name);
+		g_strfreev(name_data_splits);
 	}
 
 	g_clear_pointer(&regex, g_regex_unref);
